### PR TITLE
conformance: add conformance test for gateway addresses

### LINKED ISSUE #2035: Addresses

### PR DESCRIPTION
**Pull request by shawnh2**
**What type of PR is this?**

/kind test
/area conformance

**What this PR does / why we need it**:
add conformance tests for Addresses which are extended. the test covers single and multiple addresses.

**Which issue(s) this PR fixes**:
Fixes #2035 

**Does this PR introduce a user-facing change?**:

```release-note
No
```

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
I am not sure about this... I don't know that many implementation can support HostnameAddressType. It is extended, as is IPAddressType. I would expect most implementations actually support only one or the other?

Maybe we need it more granular with the Features? not sure, I have not dealt with this much

**Comment by shawnh2** [comment]
> I am not sure about this... I don't know that many implementation can support HostnameAddressType. It is extended, as is IPAddressType. I would expect most implementations actually support only one or the other?
> 
> Maybe we need it more granular with the Features? not sure, I have not dealt with this much

the HostnameAddressType is pretty rare, but for IPAddressType, it is supported by Envoy Gateway and Contour AFAIK. so maybe we can only foucs on IPAddressType for now.

cc @shaneutt

**Comment by shaneutt** [comment]
@shawnh2 I think what John is getting at is that we might need more specificity in the features: having a feature that's just `GatewayWithAddresses` might not be enough specificity to cover something like `HostnameAddressType`. Seems like for now at least, maybe we just have to go with a separate named feature for that.

**Comment by shawnh2** [comment]
> @shawnh2 I think what John is getting at is that we might need more specificity in the features: having a feature that's just `GatewayWithAddresses` might not be enough specificity to cover something like `HostnameAddressType`. Seems like for now at least, maybe we just have to go with a separate named feature for that.

got it! 

this PR mainly checks whether the Addresses in GatewaySpec have a corresponding entry in GatewayStatus.Addresses. so calling it `GatewayWithAddresses`  may be a little confusing, it would be nice to be more specific, like `GatewayAddresses PopulatedInStatus` etc.

File: conformance/utils/kubernetes/helpers_test.go [hunk 1]
```diff
@@ -319,3 +319,99 @@ func Test_listenersMatch(t *testing.T) {
...
+			expected: []v1beta1.GatewayAddress{
+				{
+					Type:  PtrTo(v1beta1.IPAddressType),
+					Value: "1.2.3.4",
+				},
+				{
+					Type:  PtrTo(v1beta1.HostnameAddressType),
+					Value: "test.com",
+				},
+			},
```
Comment (robscott) [inline conformance/utils/kubernetes/helpers_test.go:372]: I don't think there are many implementations that support both of these address types together. Do you know of any implementations that would pass this test?

**Comment by shaneutt** [comment]
@shawnh2 Heya, checking in :wave: 

Everything going ok with this one? Have any questions, or need any assistance moving it forward?

**Comment by shaneutt** [comment]
As this one has been inactive for a while and needs some conflict resolution, I'm going to mark it as draft temporarily as this helps us to better organize our reviews. Please feel free to move it back to ready again once you've got some time to update it, and let us know if you're stuck or need help moving this forward!

**Comment by shawnh2** [comment]
> Everything going ok with this one? Have any questions, or need any assistance moving it forward?

sorry for the late response, i will work on this issue as soon as i can.

**Comment by shaneutt** [comment]
Thanks @shawnh2: just a heads up that we consider this one a `release-blocker` for GA, so if you feel like you'll have any problem committing time to it please let us know so we can coordinate.

**Pull request closed without merging**